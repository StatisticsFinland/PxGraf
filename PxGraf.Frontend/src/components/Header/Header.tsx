import React, { useRef } from 'react';
import { useMediaQuery, Typography, Button, Box, Stack } from '@mui/material';
import styled from 'styled-components';
import LanguageSelector from 'components/LanguageSelector/LanguageSelector';
import SavedQueryFinder from 'components/SavedQueryFinder/SavedQueryFinder';
import { useNavigationContext } from 'contexts/navigationContext';
import logo from 'images/pxgraf-logo.png';
import logo_small from 'images/pxgraf-logo-small.png';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';

const Logo = styled.img`
  height: 60px;
  padding-right: 24px;

  @media (max-width: 800px) {
    padding-right: 24px;
  }
`;

const HeaderWrapper = styled(Box)`
  min-height: 75px;
`;

const LangSelectorWrapper = styled(Stack)`
  width: 20%;
`;

const MenuRowWrapper = styled(Stack)`
  padding: 8px;
  width: 100%;
  align-items: center;
  justify-content: flex-start;
`;

const LinkWrapper = styled.div`
  padding-left: 30px;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  gap: 5px;
`;

const TitleWrapper = styled.div`
  padding-left: 8px;
  min-width: 20%;
`;

/**
 * Header component displayed on top of the page in all views.
 * Contains the logo, page title, language selector, and database selector link.
 */
const Header: React.FC = () => {
    const { t } = useTranslation();
    const location = useLocation();
    const {queryId}: {queryId: string} = location?.state ? location.state : {queryId: null};
    const isNarrowScreen = useMediaQuery('(max-width: 800px)');
    const { tablePath } = useNavigationContext();
    const headerRef = useRef(null);
  
    let indexUrl = '/';
    if(tablePath?.length) {
      indexUrl = indexUrl + `?tablePath=${tablePath.join(',')}`;
    }

    // Show different top title for editor and database/table selection
    const title = location.pathname?.split("/")[1] == "editor" ? t("general.editorTitle") : t("general.selectDatabaseTitle");

    const ref = React.useRef<HTMLAnchorElement>();

    React.useEffect(() => {
      if(ref.current) ref.current.focus();
    }, [ref?.current, location.pathname])

    const focusOnContent = () => {
        const mainContentElement = document.getElementById("mainContent");
        if (mainContentElement) {
            mainContentElement.focus();
        }
    };

    return (
        <HeaderWrapper>
            <MenuRowWrapper direction="row" ref={headerRef}>
                <Button sx={{ position: 'absolute', left: '-9999px' }} href="#" onClick={(e) => { e.preventDefault(); focusOnContent(); }} ref={ref}>{t('general.contentLink')}</Button>
                <a href={indexUrl}><Logo alt={t('navbar.logoAlt')} src={isNarrowScreen ? logo_small : logo} /></a>
                <TitleWrapper>
                    <Typography variant="h1">{title}</Typography>
                </TitleWrapper>
                <LinkWrapper>
                    <Button href={indexUrl}>{t('general.databaseSelectorLink')}</Button>
                    <SavedQueryFinder oldQueryId={queryId} />
                </LinkWrapper>
                <LangSelectorWrapper direction="row-reverse" marginLeft="auto">
                    <LanguageSelector />
                </LangSelectorWrapper>
            </MenuRowWrapper>
        </HeaderWrapper>
    );
}

export default Header;