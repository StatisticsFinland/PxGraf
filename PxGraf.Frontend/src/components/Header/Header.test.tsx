import { render } from "@testing-library/react";
import Header from "./Header";

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => ({
        foo: 'bar',
    }),
    useNavigate: () => ({
        navigate: jest.fn(),
    })
}));

jest.mock('../../contexts/navigationContext', () => ({
    ...jest.requireActual('../../contexts/navigationContext'),
    useNavigationContext: () => ({tablePath: null})
}))

describe('Header component', () => {
    it('should render correctly', () => {
        const { asFragment } = render(<Header />);
        expect(asFragment()).toMatchSnapshot();
    })
});
