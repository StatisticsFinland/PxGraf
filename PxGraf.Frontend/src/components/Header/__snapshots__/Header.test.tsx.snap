// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-cGXZpB dEvZan MuiBox-root css-0"
  >
    <div
      class="MuiStack-root sc-cZztgT gSFSUc css-m69qwo-MuiStack-root"
    >
      <a
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1ycpeoo-MuiButtonBase-root-MuiButton-root"
        href="#"
        tabindex="0"
      >
        general.contentLink
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
      <a
        href="/"
      >
        <img
          alt="navbar.logoAlt"
          class="sc-jHbxoU eknJQC"
          src="test-file-stub"
        />
      </a>
      <div
        class="sc-cMHOsR exfFej"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
        >
          general.selectDatabaseTitle
        </h1>
      </div>
      <div
        class="sc-jMLmsk dSUiKC"
      >
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          href="/"
          tabindex="0"
        >
          general.databaseSelectorLink
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          savedQuery.dialogButtonTxt
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiStack-root sc-jZCRgm bUSqpf css-1os650h-MuiStack-root"
      >
        <div
          class="MuiStack-root sc-dDvxFM cpOGEN css-nen11g-MuiStack-root"
        >
          <div
            class="sc-gTrWKq iDTSPi"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 sc-jqVXSH ceSgRq css-r40f8v-MuiTypography-root"
            >
              general.uiLanguage
            </p>
            <button
              aria-label="tooltip.open: general.uiLanguage"
              class="sc-gyycJP jHfhmG"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
                data-testid="InfoIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiStack-root css-1d0f7f9-MuiStack-root"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
